applications/StructuralMechanicsApplication/custom_constitutive/linear_j2_plasticity_3d.cpp

```diff
@@ -0,0 +1,452 @@
+    return Wp_new;
+}
+
+double LinearJ2Plasticity3D::GetDeltaGamma(double NormSTrial,
+                                           const Properties& rMaterialProperties)
+{
+    const double E = rMaterialProperties[YOUNG_MODULUS];
+    const double poisson_ratio = rMaterialProperties[POISSON_RATIO];
+    const double yield_stress = rMaterialProperties[YIELD_STRESS];
+    const double theta = rMaterialProperties[REFERENCE_HARDENING_MODULUS];
+    const double hardening_modulus = rMaterialProperties[ISOTROPIC_HARDENING_MODULUS];
+    const double delta_k = rMaterialProperties[INFINITY_HARDENING_MODULUS];
+    const double hardening_exponent = rMaterialProperties[HARDENING_EXPONENT];
+    const double tolerance = 1e-6 * yield_stress;
+    const double mu = E / (2. * (1. + poisson_ratio));
+    const double sqrt_two_thirds = std::sqrt(2.0 / 3.0);
+    double dgamma = 0.0;
+    double norm_yieldfunction = 1.0;
+
+    while (norm_yieldfunction > tolerance)
+    {
+        const double k_new = GetSaturationHardening(rMaterialProperties);
+        const double kp_new = theta * hardening_modulus +
+            delta_k * (hardening_exponent * std::exp(-hardening_exponent * mAccumulatedPlasticStrain));
+        const double yieldfunction = - sqrt_two_thirds * k_new + NormSTrial - 2. * mu * dgamma;
+        const double derivative_yieldfunction = -2. * mu * (1. + kp_new / (3. * mu));
+        dgamma = dgamma - yieldfunction / derivative_yieldfunction;
+        mAccumulatedPlasticStrain = mAccumulatedPlasticStrainOld + sqrt_two_thirds * dgamma;
+        norm_yieldfunction = std::abs(yieldfunction);
+    }
+    // TODO (marcelo): handle the case when no convergence is achieved.
+    return dgamma;
+}
+
+double LinearJ2Plasticity3D::yieldFunction(const double norm_dev_stress,
+                                           const Properties& rMaterialProperties)
+{
+    const double sqrt_two_thirds = std::sqrt(2.0 / 3.0);
+    const double yield_stress = rMaterialProperties[YIELD_STRESS];
+    const double hardening_modulus = rMaterialProperties[ISOTROPIC_HARDENING_MODULUS];
+    const double theta = rMaterialProperties[REFERENCE_HARDENING_MODULUS];
+    const double delta_k = rMaterialProperties[INFINITY_HARDENING_MODULUS];
+    const double hardening_exponent = rMaterialProperties[HARDENING_EXPONENT];
+    const double k_old =
+        yield_stress + (theta * hardening_modulus * mAccumulatedPlasticStrainOld) +
+        (delta_k) * (1. - std::exp(-hardening_exponent * mAccumulatedPlasticStrainOld));
+
+    return norm_dev_stress - k_old * sqrt_two_thirds;
+}
+
+void LinearJ2Plasticity3D::CalculateElasticMatrix(Matrix &D, const Properties &props)
+{
+    const double E = props[YOUNG_MODULUS];
+    const double poisson_ratio = props[POISSON_RATIO];
+    const double lambda =
+        E * poisson_ratio / ((1. + poisson_ratio) * (1. - 2. * poisson_ratio));
+    const double mu = E / (2. + 2. * poisson_ratio);
+
+    D.clear();
```

Comment:
you don't need the D.clear, you are overwriting next. 

I would do:
if(D.size1() != 6 ...)
      D.resize(6,6,false);
D.clear(); //the effect of this is to set it to zero